gprofiler/profilers/node.py

```diff
@@ -0,0 +1,122 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+"""
+This script allows user to inject node-linux-perf module and start it to generate perf maps at runtime.
+To prepare module run npm install --prefix <location> linux-perf or clone it from https://github.com/mmarchini-oss/node-linux-perf and run npm install 
+This only works for applications, that have CommonJS as entry script.
+
+It can be also easly modified to utilize any other module by changing expression in CDP request.
+"""
+
+import asyncio
+import websockets
+import json
+import requests
+import os
+import signal
+import time
+import psutil
+import argparse
+
+EXPECTED_CDP_RESPONSE = {
+    "id": 1,
+    "result": {"result": {"type": "boolean", "value": True}},
+}
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def retry(exception, retries, waitTime):
+    def wrap(f):
+        def wrapped(*args):
+            for _ in range(0, retries):
+                try:
+                    return f(*args)
+                except exception:
+                    time.sleep(waitTime)
+
+        return wrapped
+
+    return wrap
+
+
+def __start_debugger(pid):
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def __get_debugger_url(pid):
+    process = psutil.Process(pid)
+    possible_ports = [connection.laddr.port for connection in process.connections()]
+    debugger_url = ""
+    for port in possible_ports:
+        possible_port_response = requests.get(f"http://127.0.0.1:{port}/json/list")
```

Comment:
This works only if the application is in the same network namespace, i.e not containerized.

For that we have `run_in_ns`. If you'd run this chunk of code inside `run_in_ns(["net"], target_pid=pid)` then it will work for containers as well. All networking operations you do with the target app must be done this way